library(data.table)
library(ggplot2)
library(dplyr)
library(stringr)
library(splitstackshape)

## Creates metadata table linking input file name to species name
get_true_counts <- function() {
    ids <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/id_to_genome.txt", sep=">", header=FALSE)
    ids[,c("genus", "species") := tstrsplit(V2, " ", fixed=TRUE)[2:3]]
    ids[, species:=str_replace(species, ",", "")]
    ids[, strain_count:=.N, by=.(genus, species)]
    ids[, species_full:=paste(genus, species)]
    

    ids
}


get_relevant_info <- function(input_bac, input_arc, lorikeet_output_path="", checkm_path="", parsnp_path="") {
    ## Input:
    ## input_bac - gtdbtk bacterial summary file
    ## input_arc - gtdbtk archaeal summary file
    ## lorikeet_output_path - root folder of lorikeet output
    
    ## Output:
    ## output - Table linking MAG to gtdbtk results alongside predicted strains from lorikeet
    input_bac[, species_full := tstrsplit(get("classification"), "s__", fixed=TRUE)[2]]
    input_bac[, genus := tstrsplit(species_full, " ", fixed=TRUE)[1]]
    input_bac[, genus := str_remove(genus, "_.")]
    input_bac[, species := tstrsplit(species_full, " ", fixed=TRUE)[2]]
    input_bac[, species := str_remove(species, "_.")]

    output <- input_bac[, c("user_genome", "species_full", 'genus', "species")]
    
    input_arc[, species_full := tstrsplit(get("classification"), "s__", fixed=TRUE)[2]]
    input_arc[, genus := tstrsplit(species_full, " ", fixed=TRUE)[1]]
    input_arc[, genus := str_remove(genus, "_.")]
    input_arc[, species := tstrsplit(species_full, " ", fixed=TRUE)[2]]
    input_arc[, species := str_remove(species, "_.")]

    output <- rbind(output, input_arc[, c("user_genome", "species_full", 'genus', "species")])
    output[, base_genome := tstrsplit(user_genome, "/", keep=13L)]
    
    checkm <- fread(checkm_path, skip="Bin\ Id")
    checkm[, user_genome := `Bin Id`]
    checkm[, base_genome := tstrsplit(user_genome, "/", keep=13L)]
    
    output[, 
           predicted_strains := 
           length(list.files(paste0(paste(lorikeet_output_path, base_genome, sep="/"), "/"), pattern='fna')), 
           by=user_genome]
    
    ## get tpr and fpr for called snps
    
    
    as.data.table(inner_join(output, checkm[, .(user_genome, Completeness, Contamination, `Strain heterogeneity`)]))
}

get_relevant_info_ref <- function(ids, lorikeet_output_path="", parsnp_path="") {
    ## Input:
    ## input_bac - gtdbtk bacterial summary file
    ## input_arc - gtdbtk archaeal summary file
    ## lorikeet_output_path - root folder of lorikeet output
    
    ## Output:
    ## output - Table linking MAG to gtdbtk results alongside predicted strains from lorikeet
    
    unique_ids <- unique(ids[!duplicated(species)])
    unique_ids[, user_genome:=str_replace(V1, 'Eval/Genomes/', '')]
    unique_ids[, user_genome:=str_replace(user_genome, '\\.tmp', '')]
    unique_ids[, 
           predicted_strains := 
           length(list.files(paste0(paste(lorikeet_output_path, user_genome, sep="/"), "/"), pattern='fna')), 
           by=user_genome]
    
    ## get tpr and fpr for called snps
    return(unique_ids)
}

### results function
'%!in%' <- function(x,y)!('%in%'(x,y))
'%!like%' <- function(x,y)!('%like%'(x,y))

get_results <- function(synth_input, ids) {
    ## synth_input - gtdbtk results of MAGs generated from synthetic data as generated by get_relevant_info()
    ## ids - input genome name to species name metadata table, output of get_true_counts()
    
    species_tab <- data.table(unique(full_join(synth_input, ids, by = c("species"))))
    genus <- data.table(unique(right_join(synth_input, ids[V1 %in% species_tab[is.na(get("user_genome")),]$V1], by="genus")))
    
    found <- data.table(unique(full_join(
        species_tab[!is.na(get("user_genome")) & !is.na(V1)],
        genus[!is.na(get("user_genome")) & !is.na(V1)]
        )))
    
    not_found <- ids[V1 %!in% found$V1]
    
    data.table(full_join(found, not_found))[, !c("genus.y", "V1", "V2", "species.x", "species.y")]
}

## Retreieve parsnp results
read_parsnp_vcf <- function(mag, output_folder) {
    ## mag - mag file stem
    ## output_folder - absolute path to parsnp output folder
    out <- tryCatch(
        {
            path <- paste(output_folder, mag, sep="/")
            path <- paste(path, "snps/parsnp.vcf", sep="/")
            snps <- fread(path, skip="#CHROM")
            snps %>% setnames(., old="#CHROM", new="CHROM")
            snps[, CHROM:=as.character(CHROM)]
            snps[, REF:=as.character(REF)]
            snps[, ALT:=as.character(ALT)]
            snps[, POS:=as.integer(POS)]
            snps
        },
        error = function(cond) {
            data.table(CHROM=character(), REF=character(), ALT=character(), POS=integer())
        }
    )
    return(out)
}

## Retreieve lorikeet VCF
read_lorikeet_vcf <- function(mag, output_folder) {
    ## mag - mag file stem
    ## output_folder - absolute path to lorikeet output folder
    out <- tryCatch(
        {
            path <- paste(output_folder, mag, sep="/")
            vcf_file <- paste(mag, ".vcf", sep="")
            path <- paste(path, vcf_file, sep="/")
            snps <- fread(path, sep="\t", skip="#CHROM")
            snps
            snps %>% setnames(., old="#CHROM", new="CHROM")
            snps <- snps[, snv:=ifelse(str_detect(INFO, "SNV"), 
                                         TRUE, FALSE)]
            snps <- snps[, CHROM:=tstrsplit(CHROM, "~")[2]]
            snps[, CHROM:=as.character(CHROM)]
            snps[, REF:=as.character(REF)]
            snps[, ALT:=as.character(ALT)]
            snps[, POS:=as.integer(POS)]    
            snps
        },
        error = function(cond) {
            data.table(CHROM=character(), REF=character(), ALT=character(), POS=integer(), snv=logical())
        }
    )
    return(out)
}


## Calculate confusion matrix values
confusion_matrix <- function(trues, predicted, genome_length = 3000000) {
    ## trues - set of true positive SNPs from parsnp
    ## predicted - predicted SNPs from lorikeet
    ## genome_length - the length of the input genome in base pairs
    tp_total <- nrow(trues)
    tn_total <- genome_length - tp_total
    
    tps <- nrow(inner_join(trues, predicted[snv==TRUE], by=c("CHROM", "POS", "REF", "ALT")))
    fps <- nrow(predicted[snv==TRUE]) - tps
    tns <- tn_total - fps
    fns <- tp_total - tps
    
    tpr <- tps / tp_total
    fpr <- fps / (fps + tns)
    ppv <- tps / (tps + fps)
    list(tp=tps, fp=fps, tn=tns, fn=fns, tpr=tpr, fpr=fpr, ppv=ppv)
    
}

get_confusion <- function(genome, parsnp_path, lorikeet_path) {
    parsnp <- read_parsnp_vcf(genome, parsnp_path)
    lori <- read_lorikeet_vcf(genome, lorikeet_path)
    confusion_matrix(parsnp, lori)
}

## Write vcf info to table which can be read into python
write.vars <- function(table, prefix, write=TRUE) {
    test_depths_all <- table[!is.na(CHROM), c(colnames(table) %like% "lorikeet-genome"), with=FALSE]


    test_depths <- cSplit(test_depths_all, colnames(test_depths_all), ":")

    var_depths <- test_depths[, c(colnames(test_depths) %like% "*_2"), with=FALSE]
    colnames(var_depths) <- colnames(test_depths_all)
    # var_depths[, CHROM:=test_vcf$CHROM]

    ref_depths <- test_depths[, c(colnames(test_depths) %like% "*_3"), with=FALSE]
    colnames(ref_depths) <- colnames(test_depths_all)
    # ref_depths[, CHROM:=test_vcf$CHROM]

    vars <- rbind(var_depths, ref_depths)
    if (write) {
        write.table(vars, paste0(prefix, "_test_depths.vcf"), sep="\t")
        }
    return(vars)
}

## Write labelled data for python
write.vars.labelled <- function(table, prefix, write=TRUE) {
    test_depths_all <- table[!is.na(CHROM), c(colnames(table) %like% "lorikeet-genome"), with=FALSE]
    alloc <- table[!is.na(CHROM), c(colnames(table) %like% "*seq\\.fa"), with=FALSE]
    alloc_inv <- 1 - alloc
    alloc[, bin:=strtoi(do.call(paste0, .SD), base=2)]
    alloc_inv[, bin:=strtoi(do.call(paste0, .SD), base=2)]


    test_depths <- cSplit(test_depths_all, colnames(test_depths_all), ":")

    var_depths <- test_depths[, c(colnames(test_depths) %like% "*_2"), with=FALSE]
    colnames(var_depths) <- colnames(test_depths_all)
    # var_depths[, CHROM:=test_vcf$CHROM]

    ref_depths <- test_depths[, c(colnames(test_depths) %like% "*_3"), with=FALSE]
    colnames(ref_depths) <- colnames(test_depths_all)
    # ref_depths[, CHROM:=test_vcf$CHROM]

    vars <- rbind(cbind(var_depths, alloc[, .(bin)]), cbind(ref_depths, alloc_inv[, .(bin)]))
    if (write) {
        write.table(vars, paste0(prefix, "_test_depths.vcf"), sep="\t")
    }
    return(vars)
}

remove.na <- function(DT) {
  for (i in names(DT))
    DT[is.na(get(i)), (i):=0]
}

get.all.tables <- function(mag, prefix, root, test, write=TRUE) {
    test_vcf <- read_lorikeet_vcf(mag, paste(root, test, sep="/"))
    parsnp_vcf <- read_parsnp_vcf(mag, paste(root, "parsnp/", sep="/"))
    setkey(test_vcf, CHROM, POS, REF, ALT)
    setkey(parsnp_vcf, CHROM, POS, REF, ALT)

    correct_vcf<- inner_join(test_vcf, parsnp_vcf[,], by=c("CHROM", "POS", "REF", "ALT"))[!is.na(`lorikeet-genome.sample1_R1.fq.gz`), 
                                                                                          !c("ID.x", "QUAL.x", "FILTER.x", "FORMAT.x", "INFO.x",
                                                                                            "ID.y", "QUAL.y", "FILTER.y", "FORMAT.y", "INFO.y")]

    all_vcf <- full_join(test_vcf, parsnp_vcf[,], by=c("CHROM", "POS", "REF", "ALT"))[!is.na(`lorikeet-genome.sample1_R1.fq.gz`), 
                                                                                          !c("ID.x", "QUAL.x", "FILTER.x", "FORMAT.x", "INFO.x",
                                                                                            "ID.y", "QUAL.y", "FILTER.y", "FORMAT.y", "INFO.y")]
    remove.na(all_vcf)
    # parsnp_vcf <- full_join(test_vcf, parsnp_vcf, 
    #                       by=c("CHROM", "POS", "REF", "ALT"))[is.na(`lorikeet-genome.sample1_R1.fq.gz`), !c("ID.x", "QUAL.x", "FILTER.x", "FORMAT.x", "INFO.x")]
    incorrect_vcf <- test_vcf[!parsnp_vcf]

    all <- write.vars.labelled(all_vcf, prefix, write)
    vars_correct <- write.vars.labelled(correct_vcf, paste(prefix, "correct", sep="_"), write)
    vars_incorrect <- write.vars(incorrect_vcf, paste(prefix, "incorrect", sep="_"), write)
    return(list(all=all, correct=vars_correct, incorrect=vars_incorrect))
}


# synth_15_bac <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S15D/binning/data/gtdbtk/gtdbtk.bac120.summary.tsv")
# synth_15_arc <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S15D/binning/data/gtdbtk/gtdbtk.ar122.summary.tsv")

# synth_10_bac <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S10D/binning/data/gtdbtk/gtdbtk.bac120.summary.tsv")
# synth_10_arc <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S10D/binning/data/gtdbtk/gtdbtk.ar122.summary.tsv")

# synth_05_bac <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S05D/binning/data/gtdbtk/gtdbtk.bac120.summary.tsv")
# synth_05_arc <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S05D/binning/data/gtdbtk/gtdbtk.ar122.summary.tsv")

# synth_03_bac <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S03D/binning/data/gtdbtk/gtdbtk.bac120.summary.tsv")
# synth_03_arc <- fread("/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S03D/binning/data/gtdbtk/gtdbtk.ar122.summary.tsv")


synth_15_bac <- fread("//mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S15D/combined_output/data/representatives.bac120.summary.tsv")
synth_15_arc <- fread("//mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S15D/combined_output/data/representatives.ar122.summary.tsv")

synth_10_bac <- fread("//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S10D/combined_output/data/representatives.bac120.summary.tsv")
synth_10_arc <- fread("//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S10D/combined_output/data/representatives.ar122.summary.tsv")

synth_05_bac <- fread("//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S05D/combined_output/data/representatives.bac120.summary.tsv")
synth_05_arc <- fread("//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S05D/combined_output/data/representatives.ar122.summary.tsv")

synth_03_bac <- fread("//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S03D/combined_output/data/representatives.bac120.summary.tsv")
synth_03_arc <- fread("//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S03D/combined_output/data/representatives.ar122.summary.tsv")
ids <- get_true_counts()


## With read linking
# synth_15_link <- get_relevant_info(synth_15_bac, synth_15_arc, 
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S15D/00-lorikeet_out/",
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S15D/binning/data/checkm.out"
#                                   )
# synth_10_link <- get_relevant_info(synth_10_bac, synth_10_arc, 
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S10D/00-lorikeet_out/",
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S10D/binning/data/checkm.out"
#                                   )
# synth_05_link <- get_relevant_info(synth_05_bac, synth_05_arc, 
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/00-lorikeet_out/",
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/binning/data/checkm.out"
#                                   )
# synth_03_link <- get_relevant_info(synth_03_bac, synth_03_arc, 
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S03D/00-lorikeet_out/",
#                                    "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S03D/binning/data/checkm.out"
#                                   )



## No read linking
# synth_15_no_link <- get_relevant_info(synth_15_bac, synth_15_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S15D/01-lorikeet_no_link/",
#                                        "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S15D/binning/data/checkm.out"
#                                      )
# synth_10_no_link <- get_relevant_info(synth_10_bac, synth_10_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S10D/01-lorikeet_no_link/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S10D/binning/data/checkm.out"
#                                      )
# synth_05_no_link <- get_relevant_info(synth_05_bac, synth_05_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/01-lorikeet_no_link/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/binning/data/checkm.out"
#                                      )
# synth_03_no_link <- get_relevant_info(synth_03_bac, synth_03_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S03D/01-lorikeet_no_link/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S03D/binning/data/checkm.out"
#                                      )

## Low Coverage
# synth_15_low_cov <- get_relevant_info(synth_15_bac, synth_15_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S15D/02-lorikeet_low_cov/",
#                                        "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S15D/binning/data/checkm.out"
#                                      )
# synth_10_low_cov <- get_relevant_info(synth_10_bac, synth_10_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S10D/02-lorikeet_low_cov/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S10D/binning/data/checkm.out"
#                                      )
# synth_05_low_cov <- get_relevant_info(synth_05_bac, synth_05_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/02-lorikeet_low_cov/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/binning/data/checkm.out"
#                                      )
# synth_03_low_cov <- get_relevant_info(synth_03_bac, synth_03_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S03D/02-lorikeet_low_cov/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S03D/binning/data/checkm.out"
#                                      )

## No removal

# synth_05_no_remove <- get_relevant_info(synth_05_bac, synth_05_arc, 
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/04-lorikeet_no_removal/",
#                                       "//mnt/hpccs01/scratch//microbiome/n10853499/99-STRONG_sim/Synth_G45_S05D/binning/data/checkm.out"
#                                      )

## HDBSCAN
# synth_15_hdbscan <- get_relevant_info(synth_15_bac, synth_15_arc, 
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S15D/05-lorikeet_hdbscan/",
#                                        "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S15D/binning/data/checkm.out"
#                                      )
# synth_10_hdbscan <- get_relevant_info(synth_10_bac, synth_10_arc, 
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S10D/05-lorikeet_hdbscan/",
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S10D/binning/data/checkm.out"
#                                      )
# synth_05_hdbscan <- get_relevant_info(synth_05_bac, synth_05_arc, 
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S05D/05-lorikeet_hdbscan/",
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S05D/binning/data/checkm.out"
#                                      )
# synth_03_hdbscan <- get_relevant_info(synth_03_bac, synth_03_arc, 
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S03D/05-lorikeet_hdbscan/",
#                                       "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/Synth_G45_S03D/binning/data/checkm.out"
#                                      )

## Linkage engine
synth_15_v6 <- get_relevant_info(synth_15_bac, synth_15_arc, 
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S15D/07-lorikeet_combined_output/",
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S15D/combined_output/data/representative_checkm.tsv"
                                     )
synth_15_v6 <- synth_15_v6[order(synth_15_v6$user_genome)]

synth_10_v6 <- get_relevant_info(synth_10_bac, synth_10_arc, 
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S10D/07-lorikeet_combined_output/",
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S10D/combined_output/data/representative_checkm.tsv"
                                     )
synth_10_v6 <- synth_10_v6[order(synth_10_v6$user_genome)]

synth_05_v6 <- get_relevant_info(synth_05_bac, synth_05_arc, 
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S05D/07-lorikeet_combined_output/",
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S05D/combined_output/data/representative_checkm.tsv"
                                     )
synth_05_v6 <- synth_05_v6[order(synth_05_v6$user_genome)]

synth_03_v6 <- get_relevant_info(synth_03_bac, synth_03_arc, 
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S03D/07-lorikeet_combined_output/",
                                      "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S03D/combined_output/data/representative_checkm.tsv"
                                     )
synth_03_v6 <- synth_03_v6[order(synth_03_v6$user_genome)]



## Reference genomes
ref_15 <- get_relevant_info_ref(
    ids, 
    "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S15D/08-lorikeet_representatives",
 )
ref_10 <- get_relevant_info_ref(
    ids, 
    "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S10D/08-lorikeet_representatives",
 )
ref_05 <- get_relevant_info_ref(
    ids, 
    "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S05D/08-lorikeet_representatives",
 )
ref_03 <- get_relevant_info_ref(
    ids, 
    "//mnt/hpccs01/scratch//microbiome/n10853499/02-lorikeet_testing/00-STRONG_sim/Synth_G45_S03D/08.1-lorikeet_representatives_test",
 )

## Bins to references

# synth_15_sp <- get_results(synth_15_link, ids)
# synth_10_sp <- get_results(synth_10_link, ids)
# synth_05_sp <- get_results(synth_05_link, ids)
# synth_03_sp <- get_results(synth_03_link, ids)

# write.table(synth_15_sp[, .(user_genome, V1)], "synth_15_paths.tsv", quote=FALSE, row.names=FALSE, sep="\t", col.names=FALSE)
# write.table(synth_10_sp[, .(user_genome, V1)], "synth_10_paths.tsv", quote=FALSE, row.names=FALSE, sep="\t", col.names=FALSE)
# write.table(synth_05_sp[, .(user_genome, V1)], "synth_05_paths.tsv", quote=FALSE, row.names=FALSE, sep="\t", col.names=FALSE)
# write.table(synth_03_sp[, .(user_genome, V1)], "synth_03_paths.tsv", quote=FALSE, row.names=FALSE, sep="\t", col.names=FALSE)

unique_ids <- unique(ids[!duplicated(species)])
unique_ids[,  representative:=V1]
unique_ids[, V1:=NULL]
write.table(left_join(unique_ids, ids, by=c("genus", "species"))[, .(representative, V1)], "/work/microbiome/abisko/rhys/01-projects/09-simulated_metagenomes/04-STRONG_sim/representative_paths.tsv", quote=FALSE, sep="\t", col.names=FALSE, row.names=FALSE)