{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "340f491e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.6     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.8\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.2     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m      masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\n",
      "Loading required package: plotly\n",
      "\n",
      "\n",
      "Attaching package: ‘plotly’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    last_plot\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    layout\n",
      "\n",
      "\n",
      "Loading required package: viridis\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "======================\n",
      "Welcome to heatmaply version 1.3.0\n",
      "\n",
      "Type citation('heatmaply') for how to cite the package.\n",
      "Type ?heatmaply for the main documentation.\n",
      "\n",
      "The github page is: https://github.com/talgalili/heatmaply/\n",
      "Please submit your suggestions and bug-reports at: https://github.com/talgalili/heatmaply/issues\n",
      "You may ask questions at stackoverflow, use the r and heatmaply tags: \n",
      "\t https://stackoverflow.com/questions/tagged/heatmaply\n",
      "======================\n",
      "\n",
      "\n",
      "\n",
      "---------------------\n",
      "Welcome to dendextend version 1.15.2\n",
      "Type citation('dendextend') for how to cite the package.\n",
      "\n",
      "Type browseVignettes(package = 'dendextend') for the package vignette.\n",
      "The github page is: https://github.com/talgalili/dendextend/\n",
      "\n",
      "Suggestions and bug-reports can be submitted at: https://github.com/talgalili/dendextend/issues\n",
      "You may ask questions at stackoverflow, use the r and dendextend tags: \n",
      "\t https://stackoverflow.com/questions/tagged/dendextend\n",
      "\n",
      "\tTo suppress this message use:  suppressPackageStartupMessages(library(dendextend))\n",
      "---------------------\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dendextend’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phylogram’:\n",
      "\n",
      "    prune\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:ape’:\n",
      "\n",
      "    ladderize, rotate\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    set\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cutree\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(ComplexHeatmap): there is no package called ‘ComplexHeatmap’\n",
     "output_type": "error",
     "traceback": [
      "Error in library(ComplexHeatmap): there is no package called ‘ComplexHeatmap’\nTraceback:\n",
      "1. library(ComplexHeatmap)"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "library(ggplot2)\n",
    "# library(hrbrthemes)\n",
    "library(reshape2)\n",
    "library(tidyverse)\n",
    "library(stringr)\n",
    "library(heatmaply)\n",
    "library(ape)\n",
    "library(phylogram)\n",
    "library(dendextend)\n",
    "library(ComplexHeatmap)\n",
    "library(RColorBrewer)\n",
    "library(circlize)\n",
    "theme_set(theme_bw())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1a53233",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘ComplexHeatmap’ is not available for this version of R\n",
      "\n",
      "A version of this package for your version of R might be available elsewhere,\n",
      "see the ideas at\n",
      "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "install.packages('ComplexHeatmap')\n",
    "install.packages('circlize')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "964d5067",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_ani <- function(\n",
    "    path=\"/lustre/scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/03-lorikeet_out/GCA_003139855.1_20110800_S2M_genomic/\", \n",
    "    full_metadata=full_metadata, \n",
    "    coverages\n",
    ") {\n",
    "    \"\"\"\n",
    "    Takes full path to lorikeet output folder for given genome\n",
    "    full_metadata is generated below, and coverages should be read in automatically.\n",
    "    You will just need to change the file paths to the corresponding files for you\n",
    "    \"\"\"\n",
    "    \n",
    "    suffix = tstrsplit(path, \"/\", keep=10L)\n",
    "    pop_ani <- fread(paste0(path, \"/\", suffix, \"_population_ani.tsv\"), skip=\"SampleID\",)\n",
    "    con_ani <- fread(paste0(path, \"/\", suffix, \"_consensus_ani.tsv\"), skip=\"SampleID\",)\n",
    "    sub_pop_ani <- fread(paste0(path, \"/\", suffix, \"_subpopulation_ani.tsv\"), skip=\"SampleID\",)\n",
    "    # sample_labels\n",
    "    con_ani[, SampleID:=factor(SampleID)]\n",
    "    joined_con_ani <- full_join(con_ani, full_metadata, by=c(\"SampleID\"))\n",
    "    joined_con_ani <- full_join(joined_con_ani, coverages, by=c(\"run\"=\"variable\"))\n",
    "    joined_con_ani[, sample_name_fixed := ifelse(is.na(sample_name), V1, sample_name)]\n",
    "    joined_con_ani[, log10_abundance:=scale(value)]\n",
    "    \n",
    "    pop_ani[, SampleID:=factor(SampleID)]\n",
    "    joined_pop_ani <- full_join(pop_ani, full_metadata, by=c(\"SampleID\"))\n",
    "    joined_pop_ani <- full_join(joined_pop_ani, coverages, by=c(\"run\"=\"variable\"))\n",
    "    joined_pop_ani[, sample_name_fixed := ifelse(is.na(sample_name), V1, sample_name)]\n",
    "    joined_pop_ani[, log10_abundance:=log10(value)]\n",
    "    \n",
    "    sub_pop_ani[, SampleID:=factor(SampleID)]\n",
    "    joined_sub_pop_ani <- full_join(sub_pop_ani, full_metadata, by=c(\"SampleID\"))\n",
    "    joined_sub_pop_ani <- full_join(joined_sub_pop_ani, coverages, by=c(\"run\"=\"variable\"))\n",
    "    joined_sub_pop_ani[, sample_name_fixed := ifelse(is.na(sample_name), V1, sample_name)]\n",
    "    joined_sub_pop_ani[, log10_abundance:=log10(value)]\n",
    "    \n",
    "    return(list(con=joined_con_ani, pop=joined_pop_ani, sub_pop=joined_sub_pop_ani))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a96378f",
   "metadata": {},
   "outputs": [],
   "source": [
    "header <- fread(\"/mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/03-lorikeet_out/GCA_003139855.1_20110800_S2M_genomic/GCA_003139855.1_20110800_S2M_genomic_population_ani.tsv\")\n",
    "metadata <- fread(\"/mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/sample_annotations.txt\")[run != \"-----------\"]\n",
    "metadata[, location:=ifelse(grepl(\"\\\\.NA\", sample_name, fixed=F) | startsWith(sample_name, \"20120700\"), \"Abisko\", \n",
    "                                  ifelse(startsWith(sample_name, \"G\") & taxon_name==\"peat metagenome\" | startsWith(sample_name, \"MG_\"), \"Marcell\", \n",
    "                                     ifelse(startsWith(sample_name, \"G\"), \"Loxahatchee\", \n",
    "                                            ifelse(sample_name==\"Metagenome\", \"Chersky\", \n",
    "                                                   ifelse(grepl(\"AK5YR\", sample_name, fixed=F), \"Alaska\", \n",
    "                                                          ifelse(startsWith(sample_name, \"S1071\") | startsWith(sample_name, \"BPON\"), \"Alaska\", \n",
    "                                                                 ifelse(model==\"NextSeq 500\", \"Abisko\", \n",
    "                                                                        \"Other\")))))))]\n",
    "metadata[, taxon_name:=str_replace(taxon_name, \" metagenome\", \"\")]\n",
    "metadata[, taxon_name:=str_replace(taxon_name, \"sediment\", \"sdmnt.\")]\n",
    "metadata[, type:=ifelse(location=='Abisko', ifelse(\n",
    "    grepl('_E', sample_name, fixed=F), 'Fen', ifelse(grepl('_S', sample_name, fixed=F), 'Bog', ifelse(grepl('_P', sample_name, fixed=F), 'Palsa', NA))), NA)]\n",
    "header[, V1:=str_replace(V1, \"##sample=<ID=\", \"\")]\n",
    "header[, run:=str_replace(`##source=lorikeet-v0.6.5`, \"name=04-bams/coverm-genome\\\\.\", \"\")]\n",
    "header[, `##source=lorikeet-v0.6.5`:=NULL]\n",
    "header[, run:=str_replace(run, \"_1\\\\.fastq\\\\.gz\\\\.bam>\", \"\")]\n",
    "header[, run:=str_replace(run, \"_1\\\\.fastq\\\\.bam>\", \"\")]\n",
    "\n",
    "# header$`##source=lorikeet-v0.6.1` = NULL\n",
    "full_metadata <- full_join(metadata, header, by.x=\"run\", by.y=\"run\")\n",
    "mflorens_coverage <- fread(\"//mnt/hpccs01//scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/bog_coverage_and_abundances.tsv\")\n",
    "\n",
    "mflorens_coverage_m <- as.data.table(melt(mflorens_coverage))\n",
    "mflorens_coverage_m <- mflorens_coverage_m[grepl(\"Trimmed\", variable)]\n",
    "mflorens_coverage_m[, variable:=gsub(\"_1..*\", \"\", variable)]\n",
    "mflorens_coverage_m[, variable:=gsub(\"coverm-genome\\\\.\", \"\", variable)]\n",
    "\n",
    "\n",
    "full_metadata[, SampleID:=factor(V1)]\n",
    "\n",
    "bog38 <- get_ani(\"/mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/03-lorikeet_out/GCA_003139855.1_20110800_S2M_genomic/\", full_metadata=full_metadata, coverage=mflorens_coverage_m[Genome==\"GCA_003139855.1_20110800_S2M_genomic\"])\n",
    "fen3 <- get_ani(\"/mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/03-lorikeet_out/GCA_003162175.1_20100900_E2D_genomic/\", full_metadata=full_metadata, coverage=mflorens_coverage_m[Genome==\"GCA_003162175.1_20100900_E2D_genomic\"])\n",
    "fen53 <- get_ani(\"/mnt/hpccs01/scratch/microbiome/n10853499/02-lorikeet_testing/02-mflorens_variant_test/03-lorikeet_out/GCA_003158275.1_20120600_E3D_genomic/\", full_metadata=full_metadata, coverage=mflorens_coverage_m[Genome==\"GCA_003158275.1_20120600_E3D_genomic\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "007d81bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_heatmap <- function(input_table, to_select=seq(1,208)) {\n",
    "    \"\"\"\n",
    "    Function for generating annotated ANI heatmaps\n",
    "    input table is the table generated by get_ani\n",
    "    to_select represents the range of samples to select from\n",
    "    \"\"\"\n",
    "    # Consensus heatmap\n",
    "    min_ani <- min(input_table$con[to_select, to_select + 1, with=FALSE], input_table$pop[to_select, to_select + 1, with=FALSE], input_table$sub_pop[to_select, to_select + 1, with=FALSE])\n",
    "    col_fun <- colorRamp2(c(min_ani, (1 + min_ani) / 2, 1), c(\"#2c7fb8\", \"#7fcdbb\", \"#edf8b1\"))\n",
    "    abisko_colours <- setNames(c(\"#0001FF\", \"#058000\", \"#703C1B\"), c(\"Fen\", \"Bog\", \"Palsa\"))\n",
    "#     abisko_colours <- list(\"Fen\"=\"#0001FF\", \"Bog\"=\"#058000\", \"Palsa\"=\"#703C1B\")\n",
    "    max_ref_ani_diff <- 1 - min(\n",
    "        diag(as.matrix(input_table$con[to_select, to_select + 1, with=FALSE])), \n",
    "        diag(as.matrix(input_table$pop[to_select, to_select + 1, with=FALSE])), \n",
    "        diag(as.matrix(input_table$sub_pop[to_select, to_select + 1, with=FALSE]))\n",
    "    )\n",
    "    con <- as.matrix(input_table$con[to_select, to_select + 1, with=FALSE])\n",
    "    \n",
    "    #ani difference from reference\n",
    "    con_ref_diff <- diag(con)\n",
    "    # set to NA\n",
    "    diag(con) <- NA\n",
    "    \n",
    "    rownames(con) <- input_table$con$SampleId\n",
    "    colnames(con) <- input_table$con$SampleId\n",
    "    set1_con <- RColorBrewer::brewer.pal(name=\"Set3\", n=9)\n",
    "    names(set1_con) <- sort(unique(input_table$con$taxon_name))\n",
    "    set2_con <- RColorBrewer::brewer.pal(name=\"Set3\", n=6)\n",
    "    names(set2_con) <- sort(unique(input_table$con$location))\n",
    "    \n",
    "    top_ha_con <- HeatmapAnnotation(\n",
    "        ` `=anno_barplot(1 - con_ref_diff,\n",
    "        ylim=c(0, max_ref_ani_diff))\n",
    "    )\n",
    "    bot_ha_con <- HeatmapAnnotation(\n",
    "                                 ` `=as.factor(input_table$con$location[to_select]), \n",
    "                                col=list(\n",
    "                                    ` `=set2_con\n",
    "                                    ), \n",
    "                                    show_legend = c(FALSE)\n",
    "                                )\n",
    "    row_loc <- rowAnnotation(\n",
    "                                 `Cov.`=anno_barplot(input_table$con$value[to_select]),\n",
    "                                 `Loc.`=as.factor(input_table$con$location[to_select]),\n",
    "                                 `Abisko`=as.factor(input_table$con$type[to_select]),\n",
    "                                col=list(\n",
    "                                    `Loc.`=set2_con,\n",
    "                                    `Abisko`=abisko_colours\n",
    "                                    ), show_legend = c(FALSE, FALSE, TRUE),\n",
    "                                annotation_name_rot = 90\n",
    "                                )\n",
    "\n",
    "    ha_con <- Heatmap(con, name=\"ANI\", bottom_annotation=bot_ha_con, top_annotation=top_ha_con, left_annotation=row_loc, column_title=\"conANI\", col=col_fun)\n",
    "    \n",
    "    # population\n",
    "    pop <- as.matrix(input_table$pop[to_select, to_select + 1, with=FALSE])\n",
    "    rownames(pop) <- input_table$pop$SampleId\n",
    "    colnames(pop) <- input_table$pop$SampleId\n",
    "    set1_pop <- RColorBrewer::brewer.pal(name=\"Set3\", n=9)\n",
    "    names(set1_pop) <- sort(unique(input_table$pop$taxon_name))\n",
    "    set2_pop <- RColorBrewer::brewer.pal(name=\"Set3\", n=6)\n",
    "    names(set2_pop) <- sort(unique(input_table$pop$location))\n",
    "    \n",
    "    # ani difference from reference\n",
    "    pop_ref_diff <- diag(pop)\n",
    "    # set to NA\n",
    "    diag(pop) <- NA\n",
    "    \n",
    "    top_ha_pop <- HeatmapAnnotation(` `=anno_barplot(1-pop_ref_diff,\n",
    "        ylim=c(0, max_ref_ani_diff)))\n",
    "    bot_ha_pop <- HeatmapAnnotation(` `=as.factor(input_table$pop$location[to_select]), # Variable names are just different number of spaces lol. To prevent text on concatenated plot\n",
    "\n",
    "                                col=list(\n",
    "                                    ` `=set2_pop\n",
    "                                    ), \n",
    "                                    show_legend = c(FALSE)\n",
    "                                )\n",
    "\n",
    "    ha_pop <- Heatmap(pop, name=\"ANI\", bottom_annotation=bot_ha_pop, top_annotation=top_ha_pop, column_title=\"popANI\", col=col_fun)\n",
    "    \n",
    "    # sub population\n",
    "    sub_pop <- as.matrix(input_table$sub_pop[to_select, to_select + 1, with=FALSE])\n",
    "    rownames(sub_pop) <- input_table$sub_pop$SampleId\n",
    "    colnames(sub_pop) <- input_table$sub_pop$SampleId\n",
    "    set1_sub_pop <- RColorBrewer::brewer.pal(name=\"Set3\", n=9)\n",
    "    names(set1_sub_pop) <- sort(unique(input_table$sub_pop$taxon_name))\n",
    "    set2_sub_pop <- RColorBrewer::brewer.pal(name=\"Set3\", n=6)\n",
    "    names(set2_sub_pop) <- sort(unique(input_table$sub_pop$location))\n",
    "    \n",
    "    # ani difference from reference\n",
    "    sub_pop_ref_diff <- diag(sub_pop)\n",
    "    # set to NA\n",
    "    diag(sub_pop) <- NA\n",
    "    \n",
    "    top_ha_sub_pop <- HeatmapAnnotation(`Ref. Diff`=anno_barplot(1-sub_pop_ref_diff,\n",
    "        ylim=c(0, max_ref_ani_diff)))\n",
    "    bot_ha_sub_pop <- HeatmapAnnotation(`Location`=as.factor(input_table$sub_pop$location[to_select]), \n",
    "                                        col=list(\n",
    "                                            `Location`=set2_sub_pop\n",
    "                                            ), \n",
    "                                        show_legend = c(TRUE, TRUE),\n",
    "                                        annotation_legend_param = list(\n",
    "                                            `Location` = list(direction = \"horizontal\")\n",
    "                                        ))\n",
    "\n",
    "    ha_sub_pop <- Heatmap(sub_pop, name=\"ANI\", bottom_annotation=bot_ha_sub_pop, top_annotation=top_ha_sub_pop, column_title=\"subpopANI\", col=col_fun)\n",
    "    ht_list = ha_con + ha_pop + ha_sub_pop\n",
    "    return(ht_list)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c568c647",
   "metadata": {},
   "source": [
    "# These cells print out the ANI plots to current directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd073a44",
   "metadata": {},
   "outputs": [],
   "source": [
    "bog38_list <- generate_heatmap(bog38)\n",
    "# bog38_list\n",
    "png(\"bog38_ani_maps_wide.png\",width=20,height=9,units=\"in\",res=1200)\n",
    "draw(bog38_list, merge_legend=T, column_title=\"GCA_003139855.1 ANI distributions (Bog38)\", heatmap_legend_side = \"right\", annotation_legend_side = \"bottom\")\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
